prow/crier/controller.go

```diff
@@ -160,24 +175,40 @@ func (c *Controller) processNextItem() bool {
 			return true
 		}
 
-		if c.queue.NumRequeues(key) < 5 {
-			logrus.WithError(err).WithField("prowjob", keyRaw).Error("Failed processing item, retrying")
-			c.queue.AddRateLimited(key)
-		} else {
-			logrus.WithError(err).WithField("prowjob", keyRaw).Error("Failed processing item, no more retries")
-			c.queue.Forget(key)
-			utilruntime.HandleError(err)
+		return retry(c.queue, key, err)
+	}
+
+	// initiate PrevReportState, if not exist
+	if pj.Status.PrevReportState == nil {
+		pjCopy := pj.DeepCopy()
+		pjCopy.Status.PrevReportState = map[string]v1.ProwJobState{}
+		pjCopy.Status.PrevReportState[c.reporter.GetName()] = ""
+		if _, err := c.pjclientset.Prow().ProwJobs(pj.Namespace).Patch(pjCopy); err != nil {
+			logrus.WithError(err).WithField("prowjob", keyRaw).Error("failed patching prowjob initial PrevReportState")
+			return retry(c.queue, key, err)
 		}
-		return true
 	}
 
-	if pj.Spec.Report && pj.Status.PrevReportState != pj.Status.State {
+	if pj.Spec.Report && pj.Status.PrevReportState[c.reporter.GetName()] != pj.Status.State {
 		logrus.Infof("Will report here, pj : %v, state : %s", pj.Spec.Job, pj.Status.State)
 
 		// TODO(krzyzacy): we probably should make report async as well
 		// we can also leverage this by increase number of workers.
-		if err := c.reporter.Report(pj); err == nil {
-			pj.Status.PrevReportState = pj.Status.State
+		if err := c.reporter.Report(pj); err != nil {
+			// update pj report status
+			pjCopy := pj.DeepCopy()
+			pjCopy.Status.PrevReportState[c.reporter.GetName()] = pjCopy.Status.State
+			if _, err := c.pjclientset.Prow().ProwJobs(pj.Namespace).Patch(pjCopy); err != nil {
+				// ohhhhhh no?!?! (what's the best behavior here...)
+				logrus.WithError(err).WithField("prowjob", keyRaw).Error("failed patching prowjob reported state")
+				return retry(c.queue, key, err)
+			}
+		} else {
+			// report failed, or refuse to report.. give it a retry
```

Comment:
Why does `Report()` return an error when there is no work to do?

Reply:
it needs to retry on errors?

Reply:
sorry, I mean, I have a TODO below this

Reply:
That's not an error. If there is no PubSub label, you do not want to retry. You should not return an error there

Reply:
yep, commented on the other PR